src/xwindows/keyboard.rs

```diff
@@ -0,0 +1,590 @@
+        }
+        if self.mod_is_active("Mod3") {
+            res |= KeyModifiers::SUPER;
+        }
+        //Mod2 is numlock
+        res
+    }
+
+    pub fn process_xkb_event(&self, connection: &xcb::Connection, event: &xcb::GenericEvent) {
+        let xkb_ev: &XkbGenericEvent = unsafe { xcb::cast_event(&event) };
+
+        if xkb_ev.device_id() == self.get_device_id() as u8 {
+            match xkb_ev.xkb_type() {
+                xcb::xkb::STATE_NOTIFY => {
+                    self.update_state(unsafe { xcb::cast_event(&event) });
+                }
+                xcb::xkb::MAP_NOTIFY | xcb::xkb::NEW_KEYBOARD_NOTIFY => {
+                    self.update_keymap(connection);
+                }
+                _ => {}
+            }
+        }
+    }
+    // for convenience, this fn takes &self, not &mut self
+    pub fn update_state(&self, ev: &xcb::xkb::StateNotifyEvent) {
+        self.state.borrow_mut().update_mask(
+            ev.base_mods() as xkb::ModMask,
+            ev.latched_mods() as xkb::ModMask,
+            ev.locked_mods() as xkb::ModMask,
+            ev.base_group() as xkb::LayoutIndex,
+            ev.latched_group() as xkb::LayoutIndex,
+            ev.locked_group() as xkb::LayoutIndex,
+        );
+    }
+
+    pub fn update_keymap(&self, connection: &xcb::Connection) {
+        let new_keymap = xkb::x11::keymap_new_from_device(
+            &self.context,
+            &connection,
+            self.get_device_id(),
+            xkb::KEYMAP_COMPILE_NO_FLAGS,
+        );
+        if new_keymap.get_raw_ptr() == std::ptr::null_mut() {
+            eprintln!("problem with new keymap");
+        }
+        let new_state = xkb::x11::state_new_from_device(&new_keymap, &connection, self.device_id);
+        if new_state.get_raw_ptr() == std::ptr::null_mut() {
+            eprintln!("problem with new state");
+        }
+        self.state.replace(new_state);
+        self.keymap.replace(new_keymap);
+        eprintln!("updated keymap");
+    }
+
+    pub fn get_device_id(&self) -> i32 {
+        self.device_id
+    }
+}
+
+#[cfg_attr(rustfmt, rustfmt_skip)]
+fn keysym_to_keycode(keysym: u32) -> Option<KeyCode> {
+    let res =
+        match keysym {
+            xkb::KEY_Escape =>                   KeyCode::Escape,
+            xkb::KEY_Tab =>                      KeyCode::Tab,
+          //xkb::KEY_ISO_Left_Tab =>             KeyCode::LeftTab,
+            xkb::KEY_BackSpace =>                KeyCode::Backspace,
+            xkb::KEY_Return =>                   KeyCode::Char(0xdu8 as char),
+            xkb::KEY_Insert =>                   KeyCode::Insert,
+            xkb::KEY_Delete =>                   KeyCode::Delete,
+            xkb::KEY_Clear =>                    KeyCode::Delete,
+            xkb::KEY_Pause =>                    KeyCode::Pause,
+            xkb::KEY_Print =>                    KeyCode::Print,
+          // 0x1005FF60 =>                       KeyCode::SysRq,         // hardcoded Sun SysReq
```

Comment:
There are a lot of commented sections in here.  Could you add a comment on the top of the function that explains why they are here?  For instance, is it something we might want to uncomment in the future, and if so, what we'd need to do first.  Alternatively, if we wouldn't want to uncomment them, it would be good to explain that also.   A third option is to delete the commented lines.